src/modalities/dataloader/shuffle_tokenized_data.py

```diff
@@ -0,0 +1,121 @@
+import multiprocessing
+import pickle
+import random
+from functools import partial
+from multiprocessing import Process, Queue
+from pathlib import Path
+
+from modalities.dataloader.create_packed_data import EmbeddedStreamData
+
+
+def _process_batch(batch: list[tuple[int, int]], data: bytes) -> tuple[bytes, list[int]]:
+    """
+    Process a batch of index entries to extract documents and create a new index.
+
+    Args:
+        batch (list[tuple[int, int]]): List of index entries [(start, length), ...].
+        data (bytes): Byte stream of the entire data loaded in memory.
+
+    Returns:
+        tuple[bytes, list[int]]: A tuple containing the processed data (bytes) and the list of document lengths.
+    """
+    processed_data = []
+    document_lengths = []
+
+    for start, length in batch:
+        # Access the data slice directly from the in-memory bytes
+        document = data[start : start + length]
+        processed_data.append(document)  # Already bytes
+
+        # Record the length of the document
+        document_lengths.append(length)
+
+    return b"".join(processed_data), document_lengths
+
+
+def _writer_process(output_path: Path, queue: Queue, header_data: bytes) -> None:
+    """Process to write processed data and index to the output file incrementally.
+
+    Args:
+        output_path (Path): Path to the output file.
+        queue (Queue): Queue containing processed data and document lengths.
+        header_data (bytes): Header data to write initially.
+
+    Returns:
+        None
+    """
+    with output_path.open("wb") as f:
+        # Write the header data
+        f.write(header_data)
+        current_position = 0
+        final_index = []
+
+        while True:
+            item = queue.get()
+            # Sentinel value to indicate completion
+            if item is None:
+                break
+
+            data_segment, lengths = item
+            f.write(data_segment)
+            final_index.extend((current_position, length) for length in lengths)
+            current_position += len(data_segment)
+
+        # Write the final index to the file
+        f.write(pickle.dumps(final_index))
```

Comment:
It would be nice to have one writer solution instead of two: https://github.com/Modalities/modalities/blob/b8b7bebd189af42f02a84f9cc79f8f7f3391b81d/src/modalities/dataloader/create_packed_data.py#L208
At least we should create a ticket, that we will have to come up with  a single writer implementation

Reply:
Done in https://github.com/Modalities/modalities/issues/294.